### DISCUSSION TITLE
Blending parameters with multiple stacked Tilelayers

### DISCUSSION THREAD

**Discussion by esiegel** [visgl/deck.gl#5862] [Q&A]
When rendering multiple Tilelayers what are the best or most appropriate GL Parameters to use for additive blending?

I'm currently setting these blending parameters in each of the Tilelayers:
```
              [GL.BLEND]: true,
              [GL.BLEND_DST_RGB]: GL.ONE,
              [GL.BLEND_DST_ALPHA]: GL.ONE,
              [GL.BLEND_EQUATION]: GL.FUNC_ADD,
              [GL.BLEND_SRC_RGB]: GL.ONE,
              [GL.BLEND_SRC_ALPHA]: GL.ONE,
              [GL.DEPTH_TEST]: false,
```

But I'm noticing flickering when tiles are loaded.  A tile will very suddenly be extra white before settling to correct levels.  My guess is that during tile transitions, one of the Tilelayer attributes has bad data or is potentially missing.  Definitely confused.

**NOTE**:
I think the problem has to do with the multiple overlapping tiles being displayed at the same time. I can disable this functionality in the TileLayer by using no-overlap, but this is just as alarming.

Really what I want is a way to not blend colors between tiles from the same Tileset, but continue blending between Tilesets.

thanks

**Comment by ilan-gold** [comment]
If part of the problem is the blending between the different asynchronously loading `Tilesets`, one option that comes to mind would be to either use either the `onViewportLoad` or `onTileLoad` `props` for the `TileLayer` to check when all of the tiles at a given coordinate are loaded before setting their individual `visible` parameters so that they don't blend until they are all loaded.

For within a given `Tileset` you should be able to intercept the `parameters` `prop` in your `renderSubLayers` `prop` function and just set it to `null` or whatever you want (not sure what the right answer is or what default is).

Does this help?  Maybe a codepen or something could help us understand the issue.

**Reply by esiegel** [reply]
I'll try and get a demo up sometime soon.  Using `onTileLoad` is interesting.  Maybe I could set the GL parameters to blend per Bitmap only if there isn't any tile parent or child currently visible.

**Reply by ilan-gold** [reply]
Yes that sounds like the right idea.  I'd be happy to take a look at a codepen (or equivalent way of communicating ready-to-go like a repo with a README or something) once it's ready!

**Reply by esiegel** [reply]
[Here is a repo](https://github.com/esiegel/tile_flickering) that reproduces the issue.  Had to generate some fake tile layers. 

It is based on create react app, so hopefully all you will need to run is `yarn start`.  Here is a video of the flicking I'm talking about.

https://user-images.githubusercontent.com/1390/121854331-e0051900-cca6-11eb-9076-21a080a3d3cd.mp4

Obviously, the images are contrived, but I think they show what I'm trying to accomplish.  Also, added more details in the readme.

**Reply by ilan-gold** [reply]
Hi Eric, I checked out your repo.  I see what you mean - the issue does appear to be coming from the tile refinement strategy and changing to `no-overlap` does not seem to help much in so far as there is still flickering.  I see you referenced Viv and the way we do things - I would not recommend going that route, although I do appreciate the shoutout.  It was more of a mistake, a result of my ignorance, than anything (this was one of the first things I ever coded in JS and my first introduction to graphics/WebGL).  What you have here is the right route IMO.

I think the next things to try would be:

a) Loading all the tiles at once in one layer.  This should remove the asynchronicity from the equation.  This is how Viv handles things in general.
```javascript
getTileData: async ({ x, y, z }) => {
      // early return if out of range.
      // TODO: figure out if we can avoid this bad calls
      if (x < 0 || y < 0) {
        // returning null avoids caching bad tile data
        return null;
      }

      const base = process.env.PUBLIC_URL;
      return await Promise.all(paths.map(async path => {
        const url = `${base}${path}/${z + maxDzi}/${x}_${y}.png`;
        return load(url, ImageLoader);
      }))
      
    },
````
b) Another thing to try from Viv would be a background image.  Load the lowest resolution image and scale it up so it is behind the `TileLayer` and then make sure that the `BitmapLayer` closest to the background image doesn't blend with it as you have programmed - blending happens between the two "nearest" layers I believe so basically the background image renders, the next set of tiled images renders, and then the ones on top of that blend with the next one below them so they never blend with the background image.  Does this make sense?  This combined with `no-overlap`, I think` will give you what you are looking for.  If not, we will probably have learned something, which is always nice.

**Reply by esiegel** [reply]
Loading all the tiles at once doesn't solve the issue.  In fact the flicker exists, when blending is enabled, even with a single tile instance.  This is due to the single layer blending with the refinement tile.  Using `no-overlap` fixes this issue for the single layer case, but then we see the white background as the tiles haven't loaded.  Seems like maybe a prefetch strategy would be necessary.

For idea b, I think I might be a bit confused.  I'm not sure how this would work.  The way I understand blending is that for layer that we render, and in turn for each webgl draw call, the blend equation is run for each added pixel.  In my case it adds whatever is in the buffer with whatever pixel is about to be written by the fragment shader.  So if I have three layers that I want to render I need to enforce that at most one tile from each layer is rendered over a given pixel.  By adding background images, maybe one per layer, or one composite one, and turning off blending on those layers then we have created an opaque layer that won't work.

**Reply by esiegel** [reply]
A solution I could imagine would be:
1. Draw each TileLayer into it's own framebuffer.  All blending is disabled which allows the refinement policy to work as it should
2. Render each of the framebuffers into a the screen framebuffer with blending options active.

I don't know if this is possible in deck.gl currently.

**Reply by esiegel** [reply]
Another option would be to fetch all tile channels as you suggest in a), but then do the compositing on CPU

**Reply by ilan-gold** [reply]
@esiegel  The framebuffer idea might be the way to go - the reason I thought the layers "grouped" one-next-to-another in their blending was the answer I got here: https://github.com/visgl/deck.gl/discussions/5450 and from my own experience.  That does not appear to be the case though.  

Adding the background layers does seem to alleviate greatly the flickering but the blending ends up being wrong for the reason you stated:

```javascript
new BitmapLayer({
      _imageCoordinateSystem: COORDINATE_SYSTEM.CARTESIAN,
      bounds: [0, 0 + imageHeight, 0 + imageWidth, 0],
      id: `Bitmap-${0}-bg`,
      image: makeUrl("/red_circle/red_circle_files", 0, 0, -4),
      data: null,
      parameters: {
        [GL.BLEND]: true,
        [GL.BLEND_DST_RGB]: GL.ONE,
        [GL.BLEND_EQUATION]: GL.FUNC_ADD,
        [GL.BLEND_SRC_RGB]: GL.ONE,
        [GL.DEPTH_TEST]: false,
      },
    }),
    new BitmapLayer({
      _imageCoordinateSystem: COORDINATE_SYSTEM.CARTESIAN,
      bounds: [0, 0 + imageHeight, 0 + imageWidth, 0],
      id: `Bitmap-${1}-bg`,
      image: makeUrl("/blue_circle/blue_circle_files", 0, 0, -4),
      data: null,
      parameters: {
        [GL.BLEND]: true,
        [GL.BLEND_DST_RGB]: GL.ONE,
        [GL.BLEND_EQUATION]: GL.FUNC_ADD,
        [GL.BLEND_SRC_RGB]: GL.ONE,
        [GL.DEPTH_TEST]: false,
      },
    }),
    new BitmapLayer({
      _imageCoordinateSystem: COORDINATE_SYSTEM.CARTESIAN,
      bounds: [0, 0 + imageHeight, 0 + imageWidth, 0],
      id: `Bitmap-${2}-bg`,
      image: makeUrl("/yellow_circle/yellow_circle_files", 0, 0, -4),
      data: null,
      parameters: {
        [GL.BLEND]: true,
        [GL.BLEND_DST_RGB]: GL.ONE,
        [GL.BLEND_EQUATION]: GL.FUNC_ADD,
        [GL.BLEND_SRC_RGB]: GL.ONE,
        [GL.DEPTH_TEST]: false,
      },
    createTileLayer([
      "/red_circle/red_circle_files",
      "/blue_circle/blue_circle_files",
      "/yellow_circle/yellow_circle_files",
    ]),
```

I am curious how this ends up as this is of interest to me as well.  Perhaps what I have in Viv is not actually so bad but I think I need to revisit the issue - it's possible the blending settings I was trying out just happen to play nicely.

**Reply by esiegel** [reply]
Pushed an update to branch `load_all` that fetches all layers at once and composites them via a canvas element.  Could also do the compositing in a custom Bitmap layer that supports multiple sources.  Similar to viv.
